server/src/main/java/org/apache/iotdb/db/engine/alter/TsFileRewriteExcutor.java

```diff
@@ -0,0 +1,293 @@
+import org.apache.iotdb.tsfile.file.metadata.AlignedChunkMetadata;
+import org.apache.iotdb.tsfile.file.metadata.ChunkMetadata;
+import org.apache.iotdb.tsfile.file.metadata.IChunkMetadata;
+import org.apache.iotdb.tsfile.file.metadata.enums.CompressionType;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSEncoding;
+import org.apache.iotdb.tsfile.read.TimeValuePair;
+import org.apache.iotdb.tsfile.read.TsFileAlignedSeriesReaderIterator;
+import org.apache.iotdb.tsfile.read.TsFileDeviceIterator;
+import org.apache.iotdb.tsfile.read.TsFileSequenceReader;
+import org.apache.iotdb.tsfile.read.common.Chunk;
+import org.apache.iotdb.tsfile.read.common.IBatchDataIterator;
+import org.apache.iotdb.tsfile.read.reader.IChunkReader;
+import org.apache.iotdb.tsfile.read.reader.IPointReader;
+import org.apache.iotdb.tsfile.read.reader.chunk.AlignedChunkReader;
+import org.apache.iotdb.tsfile.read.reader.chunk.ChunkReader;
+import org.apache.iotdb.tsfile.utils.Pair;
+import org.apache.iotdb.tsfile.utils.TsPrimitiveType;
+import org.apache.iotdb.tsfile.write.chunk.AlignedChunkWriterImpl;
+import org.apache.iotdb.tsfile.write.chunk.ChunkWriterImpl;
+import org.apache.iotdb.tsfile.write.schema.IMeasurementSchema;
+import org.apache.iotdb.tsfile.write.schema.MeasurementSchema;
+import org.apache.iotdb.tsfile.write.writer.TsFileIOWriter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/** This class is used to rewrite one tsfile with current encoding & compressionType. */
+public class TsFileRewriteExcutor {
+
+  private static final Logger logger = LoggerFactory.getLogger(TsFileRewriteExcutor.class);
+
+  private final TsFileResource tsFileResource;
+  private final TsFileResource targetTsFileResource;
+  private final PartialPath fullPath;
+  private final TSEncoding curEncoding;
+  private final CompressionType curCompressionType;
+  private long timePartition;
+  private boolean sequence;
+  // record the min time and max time to update the target resource
+  private long minStartTimestamp = Long.MAX_VALUE;
+  private long maxEndTimestamp = Long.MIN_VALUE;
+
+  public TsFileRewriteExcutor(
+      TsFileResource tsFileResource,
+      TsFileResource targetTsFileResource,
+      PartialPath fullPath,
+      TSEncoding curEncoding,
+      CompressionType curCompressionType,
+      long timePartition,
+      boolean sequence) {
+    this.tsFileResource = tsFileResource;
+    this.targetTsFileResource = targetTsFileResource;
+    this.fullPath = fullPath;
+    this.curEncoding = curEncoding;
+    this.curCompressionType = curCompressionType;
+    this.timePartition = timePartition;
+    this.sequence = sequence;
+  }
+
+  /** This function execute the rewrite task */
+  public void execute() throws IOException {
+
+    tsFileResource.tryReadLock();
```

Comment:
If the resource is holding the write lock by another thread, this method will not get the read lock. Use `readLock()` instead.